psl-core/src/main/java/org/linqs/psl/reasoner/dcd/DCDReasoner.java

```diff
@@ -59,82 +60,89 @@ public double optimize(TermStore baseTermStore) {
 
         termStore.initForOptimization();
 
-        // This must be called after the term store has to correct variable capacity.
-        // A reallocation can cause this array to become out-of-date.
-        float[] variableValues = termStore.getVariableValues();
-
-        double objective = -1.0;
+        long termCount = 0;
+        double objective = Double.POSITIVE_INFINITY;
         double oldObjective = Double.POSITIVE_INFINITY;
+        float[] oldVariableValues = null;
 
         if (printInitialObj && log.isTraceEnabled()) {
-            objective = computeObjective(termStore, variableValues);
-            log.trace("Iteration {} -- Objective: {}, Iteration Time: {}, Total Optimiztion Time: {}", 0, objective, 0, 0);
+            objective = computeObjective(termStore);
+            log.trace("Iteration {} -- Objective: {}, Mean Movement: {}, Iteration Time: {}, Total Optimiztion Time: {}", 0, objective, 0.0f, 0, 0);
         }
 
-        int iteration = 1;
         long totalTime = 0;
-        while (true) {
+        boolean converged = false;
+        for (int iteration = 1; iteration < (maxIterations * budget) && !converged; iteration++) {
             long start = System.currentTimeMillis();
 
+            termCount = 0;
+            objective = 0;
             for (DCDObjectiveTerm term : termStore) {
-                term.minimize(truncateEveryStep, variableValues);
+                if (oldVariableValues != null) {
+                    objective += term.evaluate(oldVariableValues);
+                } else {
+                    objective += term.evaluate(termStore.getVariableValues());
+                }
+
+                termCount++;
+                term.minimize(truncateEveryStep, termStore.getVariableValues());
             }
 
+            termStore.iterationComplete();
+
             // If we are truncating every step, then the variables are already in valid state.
             if (!truncateEveryStep) {
-                for (RandomVariableAtom variable : termStore.getVariables()) {
-                    variable.setValue(Math.max(Math.min(variable.getValue(), 1.0f), 0.0f));
+                for (int i = 0; i < termStore.getNumVariables(); i++) {
+                    termStore.getVariableValues()[i] = Math.max(0.0f, Math.min(1.0f, termStore.getVariableValues()[i]));
```

Comment:
why change this? Also does this actually store the updated value? Is it a reference being passed or a copy of the values?

Reply:
This change will update the variable values that are actually used in reasoning, the old way updated the random variable atom value which is not used until after reasoning and updated in the sync atoms call.